Refactor MappingInfo class to use XML stream parser instead of full XmlParse

JoergAtGithub opened the pull request:
Mixxx parses all XML controller mappings at every startup, but only the tiny info section is needed here.
This PR replace the parsing of the full XML file, by stream reading which is stopped once the info section is read. The info section is usually at the beginning of an XML mapping file.
On my system this PR saves ~200ms startup time.

JoergAtGithub committed:
Refactor MappingInfo class and use XML stream parser

daschuer commented on src/controllers/controllermappinginfo.cpp:
```diff
@@ -1,14 +1,25 @@
+          m_description(),
+          m_forumlink(),
+          m_wikilink(),
+          m_products() {
```
This is redundant. 

```suggestion
MappingInfo::MappingInfo(const QString& mapping_path) {
```

daschuer commented on src/controllers/controllermappinginfo.cpp:
```diff
@@ -19,102 +30,156 @@ MappingInfo::MappingInfo(const QString& mapping_path)
+        QXmlStreamReader::TokenType token = xml.readNext();
+        if (token == QXmlStreamReader::StartElement) {
+            ++xmlHierachyDepth;
+            const QString xmlElementName = xml.name().toString();
```
toString is a deep copy from a QStringView

https://github.com/qt/qtbase/blob/6d43724a30a636aa1bb719e871d23b9d0e0bc127/src/corelib/serialization/qxmlstream_p.h#L61

I think this should work: 

```suggestion
            const QStringView xmlElementName = xml.name();
```

daschuer commented on src/controllers/controllermappinginfo.cpp:
```diff
@@ -19,102 +30,156 @@ MappingInfo::MappingInfo(const QString& mapping_path)
+                    ProductInfo product;
+                    QString protocol = xmlElementAttributes
+                                               .value(QLatin1String("protocol"))
+                                               .toString();
```
Faster: 
```suggestion
                    const QStringView protocol = xmlElementAttributes
                                               .value(QStringLiteral("protocol"));
```

Swiftb0y replied:
NIT: `toString` is unnecessary afaict

daschuer commented on src/controllers/controllermappinginfo.cpp:
```diff
@@ -19,102 +30,156 @@ MappingInfo::MappingInfo(const QString& mapping_path)
-    if (!dom_forums.isNull()) {
-        m_forumlink = dom_forums.text();
-    }
+                    if (protocol == "hid") {
```
```suggestion
                    if (protocol == QStringLiteral("hid")) {
```
Same below.

daschuer reviewed (changes requested):
Thank you for this PR. I have suggested some improvements.

Swiftb0y reviewed (approved):
I agree with all the changes proposed by Daniel. LGTM to me apart from those.

JoergAtGithub committed:
Moved #include <QXmlStreamReader> to cpp file

JoergAtGithub committed:
Removed / moved unnecessary includes from controllermappinginfo.h/.cpp

JoergAtGithub committed:
Removed redundant member var initialization

JoergAtGithub committed:
Use QFileInfo directly without conversion to QString

JoergAtGithub committed:
Replaced `QLatin1String` with `QStringLiteral` for improved
performance and consistency in string comparisons and attribute
lookups. Switched from `QString` to `QStringView` for lightweight,
non-owning string handling in `xmlElementName` and `protocol`
variables, reducing memory allocations.

Swiftb0y commented on src/controllers/controllermappinginfo.cpp:
```diff
@@ -1,120 +1,177 @@
-MappingInfo::MappingInfo()
-        : m_valid(false) {
-}
+#include "util/runtimeloggingcategory.h"
```
unnecessary include?

JoergAtGithub committed:
Removed not needed include and fixed typo

Swiftb0y reviewed (approved):
Thank you. waiting for @daschuer

JoergAtGithub committed:
Added missing include

daschuer reviewed (approved):
LGTM. Thank you.
Is this ready for merge now?

JoergAtGithub commented:
Yes!
